activerecord/lib/active_record/normalization.rb

```diff
@@ -0,0 +1,151 @@
+      # _idempotent_. In other words, applying a normalization more than once
+      # should have the same result as applying it only once.
+      #
+      # Normalizations, by default, are not applied to +nil+ values. This
+      # behavior can be changed with the +:including_nil_values+ option.
+      #
+      # ==== Options
+      #
+      # * <tt>:with</tt> - The normalization to apply. May be specified as
+      #   either a callable or a Symbol which will be converted to a Proc.
+      # * <tt>:including_nil_values</tt> - Whether to apply normalizations to
+      #   +nil+ values. Defaults to +false+.
+      #
+      # ==== Examples
+      #
+      #   class User < ActiveRecord::Base
+      #     normalizes :email, with: -> email { email.strip.downcase }
+      #   end
+      #
+      #   user = User.create(email: " CRUISE-CONTROL@EXAMPLE.COM\n")
+      #   user.email                  # => "cruise-control@example.com"
+      #
+      #   user = User.find_by(email: "\tCRUISE-CONTROL@EXAMPLE.COM ")
+      #   user.email                  # => "cruise-control@example.com"
+      #   user.email_before_type_cast # => "cruise-control@example.com"
+      #
+      #   User.where(email: "\tCRUISE-CONTROL@EXAMPLE.COM ").first       # => user
+      #   User.where("email = ?", "\tCRUISE-CONTROL@EXAMPLE.COM ").first # => nil
+      def normalizes(*names, including_nil_values: false, with:)
+        with = with.to_proc if with.is_a?(Symbol)
+
+        names.each do |name|
+          attribute(name) do |cast_type|
+            NormalizedValueType.new(cast_type: cast_type, normalizer: with, normalize_nil: including_nil_values)
+          end
+        end
+
+        self.normalized_attributes += names.map(&:to_sym)
+      end
+
+      # Normalizes a given +value+ using normalizations declared for +name+.
+      #
+      # ==== Examples
+      #
+      #   class User < ActiveRecord::Base
+      #     normalizes :email, with: -> email { email.strip.downcase }
+      #   end
+      #
+      #   User.normalize(:email, " CRUISE-CONTROL@EXAMPLE.COM\n")
+      #   # => "cruise-control@example.com"
+      def normalize(name, value)
+        type_for_attribute(name).cast(value)
+      end
+    end
+
+    private
+      def normalize_changed_in_place_attributes
+        self.class.normalized_attributes.each do |name|
+          normalize_attribute(name) if attribute_changed_in_place?(name)
+        end
+      end
+
+      class NormalizedValueType < DelegateClass(ActiveModel::Type::Value) # :nodoc:
+        attr_reader :cast_type, :normalizer, :normalize_nil
+        alias :normalize_nil? :normalize_nil
+
+        def initialize(cast_type:, normalizer:, normalize_nil:)
+          @cast_type = cast_type
+          @normalizer = normalizer
+          @normalize_nil = normalize_nil
+          super(cast_type)
+        end
+
+        def normalize(value)
+          normalizer.call(value) unless value.nil? && !normalize_nil?
```

Comment:
Just a matter of taste, do you think this reads better?

```suggestion
          normalizer.call(value) if !value.nil? || normalize_nil?
```

Reply:
@MatheusRich what about :) 

```suggestion
          normalizer.call(value) if value || normalize_nil?
```

Reply:
@MatheusRich Hmm, that isn't more clear to me.  The current way reads "normalize the value unless it is nil and we don't normalize nils."  How would you describe the way you are proposing?

@psantos10 That is slightly different -- it would not call the normalizer when `value == false`.  Although I can't think of a good reason to normalize a boolean, I don't see a reason to prevent it (nor to introduce an inconsistency).